consensus/src/fallback.rs

```diff
@@ -435,95 +487,128 @@ impl Fallback {
-                for x in &b0.payload {
-                    info!("Committed B{}({})", b0.round, base64::encode(x));
+                for x in &current_block.payload {
+                    info!("Committed B{}({})", current_block.round, base64::encode(x));
                 }
+                // Cleanup the mempool.
+                self.mempool_driver.cleanup_async(&current_block).await;
             }
-
-            self.last_committed_round = b0.round;
-            debug!("Committed {:?}", b0);
-            if let Err(e) = self.commit_channel.send(b0.clone()).await {
-                warn!("Failed to send block through the commit channel: {}", e);
-            }
+            debug!("Committed {}", current_block);
+            let parent = match self.synchronizer.get_parent_block(&current_block).await? {
+                Some(b) => b,
+                None => {
+                    debug!("Commit ancestors, processing of {} suspended: missing parent", current_block.digest());
+                    break;
+                }
+            };
+            current_block = parent;
         }
-
         Ok(())
     }
 
     // #[async_recursion]
     // async fn print_chain(&mut self, block: &Block) -> ConsensusResult<()> {
+    //     if block.view < self.view {
+    //         return Ok(());
+    //     }
     //     debug!("-------------------------------------------------------- printing chain start --------------------------------------------------------");
     //     let mut current_block = block.clone();
     //     while current_block.qc != QC::genesis() {
-    //         let parent = match self.synchronizer.get_previous_block(&current_block).await? {
+    //         let parent = match self.synchronizer.get_parent_block(&current_block).await? {
     //             Some(b) => b,
     //             None => {
     //                 debug!("Processing of {} suspended: missing parent", current_block.digest());
     //                 break;
     //             }
     //         };
-    //         debug!("{:?}", current_block);
+    //         debug!("block {:?}", current_block);
     //         current_block = parent;
     //     }
-    //     debug!("{:?}", current_block);
+    //     debug!("block {:?}", current_block);
     //     debug!("-------------------------------------------------------- printing chain end --------------------------------------------------------");
     //     Ok(())
     // }
 
+    fn update_fallback_high_qc(&mut self, qc: &QC) {
+        let fallback_high_qc = match self.fallback_qcs.get(&qc.acceptor) {
+            Some(qc) => qc.clone(),
+            None => self.high_qc.clone()
+        };
+        if qc.view > fallback_high_qc.view || (qc.view == fallback_high_qc.view && qc.height > fallback_high_qc.height) {
+            self.fallback_qcs.insert(qc.acceptor, qc.clone());
+        }
+    }
+
     #[async_recursion]
     async fn process_block(&mut self, block: &Block) -> ConsensusResult<()> {
```

Comment:
This function is too long to parse one read, maybe break it into subfunctions with an if/match clause